### PR TITLE
[docs]: Document non-route files within app dir

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

A common beginner question is often around co-location of files in the `app` directory. I believe that people are simply unaware of shared routes, so I tried to explain why this project structure doesn't work.

I'm not happy with the wording on this, so some help would be appreciated. I tried to avoid terms like "colocation" because I don't believe that beginners will be familiar with the term.

# How

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: docs/pages/router/create-pages.mdx [hunk 1]
```diff
@@ -112,3 +112,20 @@ export default function Page() {
...
+<FileTree
+  files={[
+    'app/(home)/_layout.tsx',
+    ['app/(home)/profile.tsx', '/profile'],
+    'app/(modal)/_layout.tsx',
+    ['app/(modal)/profile.tsx', '/profile'],
+  ]}
+/>
+
+> **info** **Tip**: Expo recommends sorting components and hooks by "feature" and routes by "navigation pattern". While the two concepts can overlap, there is enough difference to structure them differently.
```
Comment (EvanBacon) [inline docs/pages/router/create-pages.mdx:131]: Nice, we can also document tsconfig paths as an easy way to import from the root.

File: docs/pages/router/create-pages.mdx [hunk 2]
```diff
@@ -112,3 +112,20 @@ export default function Page() {
   return <Text>Blog post: {slug}</Text>;
 }
 ```
+
+## Non-route files
+
+Your non-route files such as components and hooks cannot be placed within the `app` directory. For smaller projects may introduce seemly extra folder structure, but this rule helps prevents refactors as you adopt more advanced navigation patterns.
```
Comment (kadikraman) [inline docs/pages/router/create-pages.mdx:118]: Might I suggest rewording this? This is how I usually explain it:

> Every file and folder in the `/app` folder is either a `_layout` file, or a route in your app. This means that other files like your components, hooks, utilities etc **cannot be placed in the `/app` folder** because they're neither screens nor layout files.
> 
> A typical folder structure for a project using Expo Router might look something like this:
> 
> ```sh
> /app
> /components
> /utils
> /hooks
> app.json
> package.json
> ```

Also I'd consider adding something like this in the end since it's sure to come up:

> If you absolutely have to have one top level folder for all your code (such as a `src` folder) you can do this with a [config plugin](https://github.com/expo/universe) but we **don't recommend it** and will not accept bug reports regarding projects with custom root directories.

File: docs/pages/router/create-pages.mdx (hunk 1 above)
Comment (kadikraman) [reply docs/pages/router/create-pages.mdx:131]: An example would be really helpful here, of what sorting components and hooks by feature would look like.

File: docs/pages/router/create-pages.mdx [hunk 3]
```diff
@@ -112,3 +112,20 @@ export default function Page() {
...
+
+For example, this file structure has two different route files that both map to `/profile` and both show an identical "profile" screen. The difference between the two files are their containing `_layout`.
+
+<FileTree
```
[the remark this answers is not in the document]
Comment (Simek) [reply docs/pages/router/create-pages.mdx:120]: It's docs UI component:
* https://github.com/expo/expo/blob/4fe3a71a3f1431d06fd31e1329e1b3a140bfe6bd/docs/ui/components/FileTree/index.tsx#L19

Screenshot 2024-08-12 110740

File: docs/pages/router/create-pages.mdx [hunk 4]
```diff
@@ -112,3 +112,36 @@ export default function Page() {
...
+> In the example above both `/sign-in` and `/sign-out` are authentication routes and share components from `components/authentication`. However `/profile/[user]` is complex and renders components that belong to the authentication, profile and task features. If we place our components within the `app/` directory, its not obvious where these components belong. By separating components from routes you are creating an organized hierarchy with the philosophy that a route is a collection of reusable components, yet a component does not belong to an individual route. <br /><br />
+> Another project optimization is to utilize [path aliases](/guides/typescript/#path-aliases-optional). Path aliases keep your routes project structure agnostic, as imports within your routes are no longer dependent on our `app/` directory structure. This avoids large `import` statement changes when adding new directories (such as new groups) and prevents lengthy `import` paths.
```
Comment (EvanBacon) [inline docs/pages/router/create-pages.mdx:147]: Probably don't need all of these to be in call-out blocks.

Comment (amandeepmittal) [reply docs/pages/router/create-pages.mdx:147]: I agree with Evan's comment above. This should be part of the main section or a sub-section to "Non-route files".

Comment (amandeepmittal) [inline docs/pages/router/create-pages.mdx:147]:
```suggestion
> Another project optimization is to utilize [path aliases](/guides/typescript/#path-aliases-optional). Path aliases keep your routes project structure agnostic, as imports within your routes are no longer dependent on the **app** directory structure. This avoids large `import` statement changes when adding new directories (such as new groups) and prevents lengthy `import` paths.
```

File: docs/pages/router/create-pages.mdx [hunk 5]
```diff
@@ -112,3 +112,36 @@ export default function Page() {
...
+<FileTree
+  files={[
+    'app/(home)/_layout.tsx',
+    ['app/(home)/profile.tsx', '/profile'],
+    'app/(modal)/_layout.tsx',
+    ['app/(modal)/profile.tsx', '/profile'],
+  ]}
+/>
+
+The example above demonstrates why, if Expo Router allowed for non-route files, co-location (the technical term for keeping files within the same directory) would lead to an unstructured project. [Shared routes](/router/advanced/shared-routes) (the feature allowing you render the same route with different layouts) is common pattern in native navigation, but requires multiple directories. As you will have multiple route files, but only a single file for components co-location would force you to create arbitrary rules over which directory is the 'correct' one. Expo Router does not allow for co-location preventing this dilemma.
```
[the remark this answers is not in the document]
Comment (marklawlor) [reply docs/pages/router/create-pages.mdx]: I rewrote this section to hopefully get the point across in a simplier way

File: docs/pages/router/create-pages.mdx [hunk 6]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
...
+    'app/(modal)/_layout.tsx',
+    ['app/(modal)/profile.tsx', '/profile'],
+  ]}
+/>
+
+The example above demonstrates "why", if Expo Router allowed for non-route files, co-location (the technical term for keeping files within the same directory) would lead to an unstructured project. The example is using [shared routes](/router/advanced/shared-routes), and as such there are multiple route files for the `/profile` route. If you have non-route file `ProfileImage.tsx`, you cannot co-locate this file, as you only have one non-route file but two different route directories. To prevent this dilemma Expo Router simply does not allow for non-route file co-location.
+
+### Recommended structure
+
+Expo recommends sorting components and hooks by "feature" and routes by "navigation pattern". While the two concepts can overlap, there are enough differences to structure them differently. Additionally, routes will inevitably change. However, components generally stay within the same feature set. Keeping routes separate will greatly simplify your refactors.
```
Comment (EvanBacon) [inline docs/pages/router/create-pages.mdx:133]:
```suggestion
Expo Router recommends sorting components and hooks by "feature" and routes by "navigation pattern". While the two concepts can overlap, there are enough differences to structure them differently. Additionally, routes will inevitably change. However, components generally stay within the same feature set. Keeping routes separate will greatly simplify your refactors.
```

**Review by EvanBacon — APPROVED** [review]

File: docs/pages/router/create-pages.mdx [hunk 7]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
   return <Text>Blog post: {slug}</Text>;
 }
 ```
+
+## Non-route files
+
+Every file and sub-directory inside the **app** directory is either a `_layout` file, or a route in your app. This means that other files like your components, hooks, utilities, and so on, cannot be placed in the **app** directory because they are neither screens nor layout files.
```
Comment (amandeepmittal) [inline docs/pages/router/create-pages.mdx:118]:
```suggestion
Every file and sub-directory inside the **app** directory is either a `_layout` file or a route in your app. Other files, such as components, hooks, utilities, and so on, cannot be placed in the **app** directory because they are neither screens nor layout files.
```

There's no need to add a comma in the first line.
Let's remove the `"This means that"` since it doesn't add any benefit to the ongoing context.

File: docs/pages/router/create-pages.mdx [hunk 8]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
...
+<FileTree
+  files={[
+    'app/(home)/_layout.tsx',
+    ['app/(home)/profile.tsx', '/profile'],
+    'app/(modal)/_layout.tsx',
+    ['app/(modal)/profile.tsx', '/profile'],
+  ]}
+/>
+
+The example above demonstrates "why", if Expo Router allowed for non-route files, co-location (the technical term for keeping files within the same directory) would lead to an unstructured project. The example is using [shared routes](/router/advanced/shared-routes), and as such there are multiple route files for the `/profile` route. If you have non-route file `ProfileImage.tsx`, you cannot co-locate this file, as you only have one non-route file but two different route directories. To prevent this dilemma Expo Router simply does not allow for non-route file co-location.
```
Comment (amandeepmittal) [inline docs/pages/router/create-pages.mdx:129]:
```suggestion
The above example uses [shared routes](/router/advanced/shared-routes) for the `/profile` route inside two sub-directories. If you want to create a non-route file (for example, **ProfileImageComponent.tsx**), finding the correct sub-directory to place this file creates a dilemma since there are two route directories. This is "why" Expo Router does not allow non-route files to exist with route files inside the **app** directory because it leads to an unstructured project.
```

Regarding your earlier suggestion to use the term "co-location" and how it might add unnecessary complexity, I believe we can remove it to keep our explanation clear and concise.

File: docs/pages/router/create-pages.mdx [hunk 9]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
...
+
+### Recommended structure
+
+Expo Router recommends sorting components and hooks by "feature" and routes by "navigation pattern". While the two concepts can overlap, there are enough differences to structure them differently. Additionally, routes will inevitably change. However, components generally stay within the same feature set. Keeping routes separate will greatly simplify your refactors.
```
Comment (amandeepmittal) [inline docs/pages/router/create-pages.mdx:133]:
```suggestion
Expo Router recommends sorting components and hooks by "feature" and routes by "navigation pattern". While the two concepts can overlap, there are enough differences to structure them differently. Additionally, routes will inevitably change. However, components generally stay within the same feature set. Keeping routes separate will simplify your refactors.
```

We can remove this phrase since the sentence has enough emphasis about the explanation and providing context on simplification of refactors.

File: docs/pages/router/create-pages.mdx [hunk 10]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
...
+    'app/sign-out',
+    'app/profile',
+    'app/tasks',
+    'components/authentication',
+    'components/tasks/overview',
+    'components/profile/[user]',
+  ]}
+/>
+
+In the example above, both `/sign-in` and `/sign-out` are authentication routes and share components from the **components/authentication** file. However, `/profile/[user]` is complex and renders components that belong to the authentication, profile, and task features. If you place your components within the **app** directory, it's not obvious where these components belong. By separating components from routes, you are creating an organized hierarchy with the philosophy that a route is a collection of reusable components, yet a component does not belong to an individual route.
```
Comment (amandeepmittal) [inline docs/pages/router/create-pages.mdx:147]:
```suggestion
In the above example, both `/sign-in` and `/sign-out` are authentication routes and share components from the **components/authentication** file. However, the `/profile/[user]` route is complex and renders components that belong to the authentication, profile, and task features. If you place your components within the **app** directory, it's not obvious where these components belong. By separating components from routes, you create an organized hierarchy with the philosophy that a route is a collection of reusable components, and a component does not belong to an individual route.
```

File: docs/pages/router/create-pages.mdx [hunk 11]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
...
+
+Another project optimization is to utilize [path aliases](/guides/typescript/#path-aliases-optional). Path aliases keep your routes project structure agnostic, as imports within your routes are no longer dependent on the **app** directory structure. This avoids large `import` statement changes when adding new directories (such as new groups) and prevents lengthy `import` paths.
```
Comment (amandeepmittal) [inline docs/pages/router/create-pages.mdx:149]:
```suggestion
Another project optimization is to utilize [path aliases](/guides/typescript/#path-aliases-optional). They keep your routes project structure agnostic, as imports within your routes no longer depend on the **app** directory structure. This avoids large `import` statement changes when adding new directories (such as new groups) and prevents lengthy `import` paths.
```

File: docs/pages/router/create-pages.mdx [hunk 12]
```diff
@@ -112,3 +112,38 @@ export default function Page() {
...
+  files={[
+    'app/sign-in',
+    'app/sign-out',
+    'app/profile',
+    'app/tasks',
+    'components/authentication',
+    'components/tasks/overview',
+    'components/profile/[user]',
+  ]}
+/>
```
Comment (webda2l) [inline docs/pages/router/create-pages.mdx:145]:
```
<FileTree
  files={[
    'app/sign-in',
    'app/sign-out',
    'app/profile/[user]',
    'app/tasks',
    'components/authentication',
    'components/tasks/overview',
    'components/profile',
  ]}
/>
```

is the one expected no?
